src/main/java/net/dries007/tfc/world/surface/SoilSurfaceState.java

```diff
@@ -46,29 +46,40 @@ public BlockState getState(SurfaceBuilderContext context)
         final float rainfall = context.rainfall();
         final ChunkData data = context.getChunkData();
 
+        final float forest = data.getAdjustedForestDensity();
         if (rainfall < RAINFALL_SANDY)
         {
             if (rainfall > RAINFALL_SAND_SANDY_MIX)
             {
                 // Sandy
                 return soil(SoilBlockType.Variant.SANDY_LOAM);
             }
-            else if (rainfall > RAINFALL_SAND)
+            else
             {
-                // Sandy - Sand Transition Zone
-                float noise = patchNoise.noise(pos.getX(), pos.getZ());
-                if (noise > 0.2f && data.getAdjustedForestDensity() > 0.5f)
+                final double slope = context.getSlope();
+                final float weirdness = data.getForestWeirdness();
+                if ((weirdness < 0.18f || (weirdness < 0.22f && (weirdness - context.random().nextFloat() * 0.5f < 0f))) && slope < 2 && (slope < 1 || context.random().nextBoolean()))
```

Comment:
There's no comment here, what is this zone trying to do? The slope values seem very questionable to me.